sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -655,6 +655,8 @@ class Analyzer(
         resolved
       } else {
         plan match {
+          case g: Generate =>
+            if (g.join) resolveExpressionRecursively(resolved, g.child) else resolved
```

Comment:
The `join` is not specified by user explicitly. If we allow `sort` refer to the column that is not in project list, I think we should also allow it even `join` is false, it means we need to change `join` from `false` to `true` to make it works.